### ISSUE TITLE
RequestValidator generating invalid comparison hash.

### ISSUE THREAD

**Issue by ellisio** [twilio/twilio-php#436]
When implementing the security check for webhooks on Laravel 5.5 the signature being generated on the webhook request does not match. 

**Version:** 5.7.3
**API Subdomain (api/taskrouter/ip_messaging):** Security

### Code Snippet
```php
protected function validTwilioRequest()
{
    $validator = new RequestValidator(env('TWILIO_AUTH_TOKEN'));

    return $validator->validate(
        $this->header('X-Twilio-Signature'),
        $this->fullUrl(),
        $this->all()
    );
}
```

### Exception Log
```
[2017-07-04 05:12:05] local.INFO: array (
  'SmsSid' => 'SMe58b190230fa463d9c23638ff6858de4',
  'SmsStatus' => 'delivered',
  'MessageStatus' => 'delivered',
  'To' => '+1480<redacted>',
  'MessageSid' => 'SMe58b190230fa463d9c23638ff6858de4',
  'AccountSid' => 'AC1f45ea18c7b5317c06b1f7d445d597bd',
  'From' => '+1720<redacted>',
  'ApiVersion' => '2010-04-01',
) (REQUEST PARAMS "$this->all()")
[2017-07-04 05:12:05] local.INFO: https://ellis.ngrok.io/webhook/twilio/status (FULL URL "$this->fullUrl()")
[2017-07-04 05:12:05] local.INFO: GiGBaTH7lUw4sMnn6RRGLUeaq3A= (X-Twilio-Signature)
[2017-07-04 05:12:05] local.INFO: C5ed4AQZ7rXyFdrRVo7xFq+h94Q= (GENERATED SIG)
```

### Steps to Reproduce
1. Generate a request with `'statusCallback' => url('/webhook/twilio/status')`
2. Observe the inbound request to the webhook.
3. Logged the received signature with the generate signature, verified they do not match.

**Comment by jamesgraham** [comment]
Can you share your `RequestValidator` - we have a similar function on L5.4 working ok

**Comment by ellisio** [comment]
@jamesgraham This is the baked in RequestValidator in the SDK: https://github.com/twilio/twilio-php/blob/master/Twilio/Security/RequestValidator.php

**Comment by johnpaulmedina** [comment]
I too have this issue when using the RequestValidator. I believe its due to GET and POST issues. I can not get the request signatures to match. Im on L5.2 but I had this working on L5.1

**Comment by imthepitts** [comment]
@ellisio validation failures are generally due to an input problem. Can you try with vanilla PHP to see if Laravel is interfering?

**Comment by ellisio** [comment]
@imthepitts Laravel isn't modifying the data in any way that would effect this request. I am using Ngrok for local development and the data you see in my exception log from `$request->all()` is the same data on the POST body that Ngrok displays on the request.

It's also worth mentioning that Laravel Request extends Symfony's Request which creates a request object using `$_POST` passed into the Symfony ParameterBag object.

> TL;DR: The request data in Laravel's request object is the raw $_POST data from the Nginx request, nothing is modified.

**Comment by ellisio** [comment]
By the way, here is the URL that is being generated by `RequestValidator` prior to being used by `hash_hmac`:

```
https://ellis.ngrok.io/webhook/twilio/statusaccountsidac1f45ea18c7b5317c06b1f7d445d597bdapiversion2010-04-01from+17205360428messagesidsm43cb992c768b40ceac387f1487cc0db9messagestatusdeliveredsmssidsm43cb992c768b40ceac387f1487cc0db9smsstatusdeliveredto+14802389944
```

Generated by these lines. https://github.com/twilio/twilio-php/blob/master/Twilio/Security/RequestValidator.php#L17-L22

**Comment by imthepitts** [comment]
@ellisio thanks for the sample URL. I noticed all the POST parameters and values appended to your URL have been changed to lower case. Twilio callbacks use CamelCase for parameter names, and the validator doesn't change the case of the parameters or values. 

I ran a basic test myself and generated this URL (redacted):

`https://de7b4b12.ngrok.io/projects/php/signature.phpAccountSidACxxxxApiVersion2010-04-01From+1xxxxxMessageSidSM58b6377df026490aaf00a7496ae36c7fMessageStatussentSmsSidSMxxxxSmsStatussentTo+1xxxx`

This validates correctly right out of the box. Can you check where in your request pipeline the parameter case is getting changed?

**Comment by ellisio** [comment]
Hey @imthepitts, I cleared our `vendor` directory and re-ran `composer install` to ensure I had a clean copy of the SDK. After doing so, this is the data being passed into `computeSignature()`.

`$url`: `https://ellis.ngrok.io/webhook/twilio/sms`
`$data`:
```
Array
(
    [ToCountry] => US
    [ToState] => CO
    [SmsMessageSid] => SM904c5ab4d667f2fd03cdfbb4a08e1149
    [NumMedia] => 0
    [ToCity] => DENVER
    [FromZip] => 85021
    [SmsSid] => SM904c5ab4d667f2fd03cdfbb4a08e1149
    [FromState] => AZ
    [SmsStatus] => received
    [FromCity] => PHOENIX
    [Body] => Hi
    [FromCountry] => US
    [To] => +1{redacted}
    [ToZip] => 80014
    [NumSegments] => 1
    [MessageSid] => SM904c5ab4d667f2fd03cdfbb4a08e1149
    [AccountSid] => AC1f45ea18c7b5317c06b1f7d445d597bd
    [From] => +1{redacted}
    [ApiVersion] => 2010-04-01
)
```

This generates the following `$url` prior to HMAC:

```
https://ellis.ngrok.io/webhook/twilio/smsAccountSidAC1f45ea18c7b5317c06b1f7d445d597bdApiVersion2010-04-01BodyHiFrom+1{redacted}FromCityPHOENIXFromCountryUSFromStateAZFromZip85021MessageSidSM904c5ab4d667f2fd03cdfbb4a08e1149NumMedia0NumSegments1SmsMessageSidSM904c5ab4d667f2fd03cdfbb4a08e1149SmsSidSM904c5ab4d667f2fd03cdfbb4a08e1149SmsStatusreceivedTo+1{redacted}ToCityDENVERToCountryUSToStateCOToZip80014
```

As you can see, no longer `strtolower`. However, the validator is still returning false.

I notice that the validator takes the `TWILIO_AUTH_TOKEN` value that I have set in my .env. Could the issue possibly be caused by the fact that the Inbound messages is through a _Sub-Account_ and not the _Master Account_?

**Comment by imthepitts** [comment]
@ellisio, the Master Account's Auth Token is only used to sign webhook requests on the Master Account itself. So if you're using the Master Auth Token to sign webhook requests on a Sub-Account, signature validation would be expected to fail. Please try the Sub-Account Auth Token and let us know if you continue having trouble.

All the Best,
Zack

**Comment by ellisio** [comment]
@imthepitts That indeed did the trick. We now have request validation working via the following:

```
$account = $this->twilio($this->get('AccountSid'))->fetch();

$validator = new RequestValidator($account->authToken);

return $validator->validate(
    $this->header('X-Twilio-Signature'),
    $this->fullUrl(),
    $this->all()
);
```

**Comment by pfeiferchristopher** [comment]
So I am having issues with the Validator's hash not matching.

`$url` = `https://app.salesmessage.com/webhook/twilio/sms`

`$data`
```json
{
  "ToCountry":"US",
  "ToState":"UT",
  "SmsMessageSid":"SMcb0e2a8c3d092e0439dd0fcc3e237a48",
  "NumMedia":"0", 
  "ToCity":"BOUNTIFUL",
  "FromZip":null,
  "SmsSid":"SMcb0e2a8c3d092e0439dd0fcc3e237a48",
  "FromState":"MI",
  "SmsStatus":"received",
  "FromCity":null,
  "Body":"Okay I will look",
  "FromCountry":"US",
  "To":"+1{redacted}",
  "ToZip":"84010",
  "NumSegments":"1",
  "MessageSid":"SMcb0e2a8c3d092e0439dd0fcc3e237a48",
  "AccountSid":"AC45c77cada02af0d714894c6ed21635c2",
  "From":"+1{redacted}",
  "ApiVersion":"2010-04-01"
}
```

The following Full URL `$url` was generated:
```
https://app.salesmessage.com/webhook/twilio/smsAccountSidAC45c77cada02af0d714894c6ed21635c2ApiVersion2010-04-01Body{redacted}From+1{redacted}FromCityFromCountryUSFromStateMIFromZipMessageSidSMcb0e2a8c3d092e0439dd0fcc3e237a48NumMedia0NumSegments1SmsMessageSidSMcb0e2a8c3d092e0439dd0fcc3e237a48SmsSidSMcb0e2a8c3d092e0439dd0fcc3e237a48SmsStatusreceivedTo+1{redacted}ToCityBOUNTIFULToCountryUSToStateUTToZip84010
```

What was happen previously, was the payload also had a `AddOns` object as part of the payload that `computedSignature` was not properly turning into a string.

Anyone see anything wrong here?

**Comment by ellisio** [comment]
@imthepitts This issue is creeping up again. Using the same code as above, with the `5.16.0` SDK, the validator is failing.

```
{
   "url":"https://textloop.ngrok.io/twilio",
   "signature":"+JeNtnSv0u8+VgWugxqRCxM+2Z4=",
   "twiliosignature":"j3jDZr1iLa+ONOEz7vIVuQLapC4=",
   "data":{
      "ToCountry":"US",
      "ToState":null,
      "SmsMessageSid":"SM3e8f09f06b1b7db946dfdc443e1a6f66",
      "NumMedia":"0",
      "ToCity":null,
      "FromZip":"85021",
      "SmsSid":"SM3e8f09f06b1b7db946dfdc443e1a6f66",
      "FromState":"AZ",
      "SmsStatus":"received",
      "FromCity":"PHOENIX",
      "Body":"Test",
      "FromCountry":"US",
      "To":"+1844redacted",
      "ToZip":null,
      "NumSegments":"1",
      "MessageSid":"SM3e8f09f06b1b7db946dfdc443e1a6f66",
      "AccountSid":"ACf04e706b844a8255bc1b6255d594e097",
      "From":"+1480redacted",
      "ApiVersion":"2010-04-01"
   }
}
```

The validator generates the following URL:

```
https://textloop.ngrok.io/twilioAccountSidACf04e706b844a8255bc1b6255d594e097ApiVersion2010-04-01BodyTestFrom+1480redactedFromCityPHOENIXFromCountryUSFromStateAZFromZip85021MessageSidSM3e8f09f06b1b7db946dfdc443e1a6f66NumMedia0NumSegments1SmsMessageSidSM3e8f09f06b1b7db946dfdc443e1a6f66SmsSidSM3e8f09f06b1b7db946dfdc443e1a6f66SmsStatusreceivedTo+1844redactedToCityToCountryUSToStateToZip
```

As you can see, the two signatures are incorrect again.

**Comment by ellisio** [comment]
Here is the sorted request data (from ksort in the validator):

```
[2017-11-28 22:09:09] local.INFO: Array
(
    [AccountSid] => ACf04e706b844a8255bc1b6255d594e097
    [ApiVersion] => 2010-04-01
    [Body] => Test
    [From] => +1480redacted
    [FromCity] => PHOENIX
    [FromCountry] => US
    [FromState] => AZ
    [FromZip] => 85021
    [MessageSid] => SM3e8f09f06b1b7db946dfdc443e1a6f66
    [NumMedia] => 0
    [NumSegments] => 1
    [SmsMessageSid] => SM3e8f09f06b1b7db946dfdc443e1a6f66
    [SmsSid] => SM3e8f09f06b1b7db946dfdc443e1a6f66
    [SmsStatus] => received
    [To] => +1844redacted
    [ToCity] =>
    [ToCountry] => US
    [ToState] =>
    [ToZip] =>
)
```

**Comment by pfeiferchristopher** [comment]
I'm getting the same thing, this occurred just a few moments ago.

```json
{
    "url":"https://app.salesmessage.com/webhook/twilio/sms",
    "signature":"oNaOL6lMboRwKn9MNbN6PGm/EnM=",
    "twiliosignature":"aGrEuo/xj9hQes7NeU8wPASFyWk=",
    "data":{
        "ToCountry":"US",
        "ToState":"CA",
        "SmsMessageSid":"SM105{redacted}",
        "NumMedia":"0",
        "ToCity":"MOUNTAIN VIEW",
        "FromZip":"95481",
        "SmsSid":"SM105{redacted}",
        "FromState":"CA",
        "SmsStatus":"received",
        "FromCity":"EUREKA",
        "Body":"{redacted}",
        "FromCountry":"US",
        "To":"+1{redacted}",
        "ToZip":"94040",
        "AddOns":{
            "status":"successful",
            "message":null,
            "code":null,
            "results": {}
        },            
        "NumSegments":"1",
        "MessageSid":"SM105{redacted}",
        "AccountSid":"AC459{redacted}",
        "From":"+1{redacted}",
        "ApiVersion":"2010-04-01"
    }
}
```

**Comment by eaharonoff** [comment]
any clarification on how to address this issue? I ran into this as well and am having trouble debugging it. Using L5.3/PHP 7.1.

**Comment by childish-sambino** [comment]
Closing since this issue is fairly old and there have been changes made to the validator since. Feel free to re-open or create a separate issue if this is still on-going.
